pkg/controller/common/util.go

```diff
@@ -1887,25 +1898,74 @@ func GetSnapshotClassForSmartClone(dvName string, targetPvcStorageClassName *str
 		return "", nil
 	}
 
-	// List the snapshot classes
-	scs := &snapshotv1.VolumeSnapshotClassList{}
-	if err := client.List(context.TODO(), scs); err != nil {
-		logger.Info("Cannot list snapshot classes, falling back to host assisted clone")
+	vscName, err := GetVolumeSnapshotClass(context.TODO(), client, pvc, targetStorageClass.Provisioner, snapshotClassName, logger, recorder)
+	if err != nil {
 		return "", err
 	}
-	for _, snapshotClass := range scs.Items {
-		// Validate association between snapshot class and storage class
-		if snapshotClass.Driver == targetStorageClass.Provisioner {
+	if vscName != nil {
+		if pvc != nil {
 			logger.Info("smart-clone is applicable for datavolume", "datavolume",
-				dvName, "snapshot class", snapshotClass.Name)
-			return snapshotClass.Name, nil
+				pvc.Name, "snapshot class", *vscName)
 		}
+		return *vscName, nil
 	}
 
 	logger.Info("Could not match snapshotter with storage class, falling back to host assisted clone")
 	return "", nil
 }
 
+// GetVolumeSnapshotClass looks up the snapshot class based on the driver and an optional specific name
```

Comment:
Should we add a comment for the defined fallback behavior?

Reply:
sure, added.